USB/IP simulation

szszszsz opened the pull request:
This patch allows to use USB/IP as a mean to simulate device connection
to the OS.

Platform and storage implementation taken from the Trussed tutorial:
- https://github.com/trussed-dev/trussed-totp-pc-tutorial

Remarks:
- FIDO app only (to be extended with Admin and Provision apps);
- works with Chromium and pynitrokey (with patched fido2.hid module);
- written length returns "1", which confuses client HID applications
  (Chromium shows error in logs, but ignores it; pynitrokey fails);
- currently not possible to set the FIDO certificate, thus x5c response
  is empty;
- does not work with Firefox at the moment;
- requires multiple `usbip attach` calls to make it work [1];
- Makefile helper provided for convenience:
    - `make` to start;
    - `make stop` to kill application;
- simple environment to build provided via a Dockerfile.

Tested using Linux Fedora 33 with kernel 5.13.19-100.fc33.x86_64 with:
- Chromium Version 99.0.4844.51 (Official Build) Fedora Project (64-bit)
    - works on [3], and [2] (but only RK);
- Firefox 98.0
    - does not work due to lack of the CTAP1 support in this patch;
on [2][3].

[1] https://github.com/Sawchord/usbip-device#known-bugs
[2] https://webauthn.io/
[3] https://webauthn.bin.coffee/

Connected:
- https://github.com/Nitrokey/nitrokey-3-firmware/pull/32

szszszsz commented:
Requesting review for this PR. Potential things to check:
- [ ] PID/VID - currently set to Nitrokey 3's values AFAIR -> can we have a configuration file for that?
- [ ] dependencies (potentially out of date)

szszszsz commented:
Now it builds and runs. Dependencies updated. PID/VID configurable from the CLI -> defaults can be changed to Solo2.

robin-nitrokey commented on runners/pc-usbip/Makefile:
```diff
@@ -0,0 +1,45 @@
+	sudo usbip detach -p "00"
+	killall $(APPNAME)
+
+.PHONY: stop
```
s/stop/setup-fedora/

robin-nitrokey commented on runners/pc-usbip/README.md:
```diff
@@ -0,0 +1,49 @@
+# USB/IP Simulation
+
+This runner allows using USB/IP as a mean to simulate device connection
```
s/mean/means/

robin-nitrokey commented on runners/pc-usbip/README.md:
````diff
@@ -0,0 +1,49 @@
+make stop
+```
+
+Warning: in some cases' simulation can sometimes cause kernel faults, which makes the system it is running unstable. 
````
s/'//

robin-nitrokey commented on runners/pc-usbip/src/bin/main.rs:
```diff
@@ -0,0 +1,84 @@
+    let args = Args::parse();
+
+    log::info!("Initializing Trussed");
+    let state_file: String = args.state_file;
```
type annotation should not be necessary here

robin-nitrokey commented on runners/pc-usbip/src/platform.rs:
```diff
@@ -0,0 +1,77 @@
+    }
+}
+
+impl trussed::platform::UserInterface for UserInterface
+{
```
remove linebreak

robin-nitrokey commented on runners/pc-usbip/Makefile:
```diff
@@ -0,0 +1,45 @@
+.PHONY: start-sim
+start-sim: $(APPNAME)
+	-$(MAKE) stop
+	cp $(APPPATH) ./$(APPNAME) -v
+	env RUST_LOG=debug ./$(APPNAME) &
```
Did you consider using `cargo run` instead of copying around the binary and executing it directly?  If you don’t want to use `cargo run`, taking `CARGO_TARGET_DIR` into account when setting `APPPATH` (or making it possible to override `APPPATH`) would help users using a non-default target directory (for example me :)).

robin-nitrokey commented on runners/pc-usbip/README.md:
```diff
@@ -0,0 +1,49 @@
+
+Remarks:
+- At the moment FIDO app only (to be extended with Admin and Provision apps);
+- Works with Chromium and pynitrokey (with a patched fido2.hid module);
```
Could you please add some details on the required patch?

robin-nitrokey replied:
Looks like you just have to replace `raise OSError("failed to write entire packet")` with `pass` in `FileCtapHidConnection.write_packet` in `fido2`’s `hid/base.py`.

https://github.com/Yubico/python-fido2/blob/dd5662ca7b38ce6b30eb6c916c92bc75006424e7/fido2/hid/base.py#L38

robin-nitrokey reviewed (approved):
Great work, thank you!  Tested with Chromium 100.0.4896.60  and webauthn.bin.coffee on Debian stable.  I’ve added some comments with minor suggestions.

robin-nitrokey commented:
I’ve added support for multiple apps, added the admin app and added command line options for FIDO key and cert provisioning on our [usb-ip-apps branch](https://github.com/Nitrokey/nitrokey-3-firmware/tree/usb-ip-apps).  With this, `nitropy nk3 test` passes.

Next steps for me:
- [ ] add command-line options for admin app settings
- [ ] read defaults for the command-line arguments from a configuration file

szszszsz commented:
I have addressed some of the review comments, but for some reason the commit list / whole PR changes list were not updated here. Will revisit this tomorrow.
Sources were pushed to the new place, and thus this PR will be soon closed. The development will be continued here:
- https://github.com/trussed-dev/pc-usbip-runner

szszszsz commented:
Closing as done
